File: ros1bookcode-master-rw/chapt10/chapt10_ws/src/learn_queue_latch_py/learn_queue_latch_py/queue_latch_demo.py
#!/usr/bin/env python3
import rospy
from nav_msgs.msg import Odometry

class OdomPublisherSubscriber:
    def __init__(self):
        self.odom_publisher = rospy.Publisher('odom', Odometry, queue_size=10, latch=True)
        self.odom_subscriber = rospy.Subscriber('odom', Odometry, self.odom_callback, queue_size=10)
        self.timer = rospy.Timer(rospy.Duration(1.0), self.timer_callback)
    def odom_callback(self, _):
        rospy.loginfo('收到里程计消息')
    def timer_callback(self, _):
        self.odom_publisher.publish(Odometry())

def main():
    rospy.init_node('queue_latch_demo')
    OdomPublisherSubscriber()
    rospy.spin()
if __name__ == '__main__':
    main()

File: ros1bookcode-master-rw/chapt10/chapt10_ws/src/learn_queue_latch_cpp/src/queue_latch_demo.cpp
#include <ros/ros.h>
#include <nav_msgs/Odometry.h>

class OdomPublisherSubscriber {
public:
  OdomPublisherSubscriber() {
    odom_publisher_ = nh_.advertise<nav_msgs::Odometry>("odom", 10, true);
    odom_subscription_ = nh_.subscribe("odom", 10, &OdomPublisherSubscriber::callback, this);
    timer_ = nh_.createTimer(ros::Duration(1.0), &OdomPublisherSubscriber::timerCallback, this);
  }
private:
  void callback(const nav_msgs::Odometry::ConstPtr&) { ROS_INFO("收到里程计消息"); }
  void timerCallback(const ros::TimerEvent&) { odom_publisher_.publish(nav_msgs::Odometry()); }
  ros::NodeHandle nh_;
  ros::Publisher odom_publisher_;
  ros::Subscriber odom_subscription_;
  ros::Timer timer_;
};

int main(int argc, char* argv[]) {
  ros::init(argc, argv, "queue_latch_demo");
  OdomPublisherSubscriber node;
  ros::spin();
  return 0;
}
